src/data/battler-tags.ts

```diff
@@ -92,6 +92,110 @@ export interface TerrainBattlerTag {
   terrainTypes: TerrainType[];
 }
 
+/**
+ * Base class for tags that disable moves. Descendants can override {@linkcode moveIsDisabled} to disable moves that
+ * match a condition. A disabled move gets cancelled before it is used. Players and enemies should not be allowed
+ * to select disabled moves.
+ */
+export abstract class DisablingBattlerTag extends BattlerTag {
+  public abstract moveIsDisabled(move: Moves): boolean;
+
+  constructor(tagType: BattlerTagType, turnCount: integer, sourceMove?: Moves, sourceId?: integer) {
+    super(tagType, [ BattlerTagLapseType.PRE_MOVE, BattlerTagLapseType.TURN_END ], turnCount, sourceMove, sourceId);
+  }
+
+  lapse(pokemon: Pokemon, lapseType: BattlerTagLapseType): boolean {
+    if (lapseType === BattlerTagLapseType.PRE_MOVE) {
+      // Cancel the affected pokemon's selected move
+      const phase = pokemon.scene.getCurrentPhase() as MovePhase;
+      const move = phase.move;
+
+      if (!this.moveIsDisabled(move.moveId)) {
+        return true;
+      }
+
+      pokemon.scene.queueMessage(this.interruptedText(pokemon, move.moveId));
+      phase.fail();
+    } else if (lapseType === BattlerTagLapseType.TURN_END) {
+      // On turn end, subtract from lifetime and remove this tag if 0
+      if (!super.lapse(pokemon, lapseType)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
```

Comment:
```suggestion
  lapse(pokemon: Pokemon, lapseType: BattlerTagLapseType): boolean {
    if (lapseType === BattlerTagLapseType.PRE_MOVE) {
      // Cancel the affected pokemon's selected move
      const phase = pokemon.scene.getCurrentPhase() as MovePhase;
      const move = phase.move;

      if (this.moveIsDisabled(move.moveId)) {
        pokemon.scene.queueMessage(this.interruptedText(pokemon, move.moveId));
        phase.fail();
      }
      return true;
    }
    // On turn end, subtract from lifetime and remove this tag if 0
    return super.lapse(pokemon, lapseType);
  }
```
Some style changes, should still function the same.